File: src/syslog.ts
import { SyslogOptions, SyslogFacility, SyslogLevel, SyslogOption } from './types';

/**
 * Native module loader with fallback
 * @internal
 */
declare const require: (id: string) => any;
declare const __dirname: string;

function loadNativeModule() {
  try {
    // Try to load the precompiled binary first
    return require('../lib/binding/syslog_native.node');
  } catch (error) {
    try {
      // Fallback to the source module
      return require('node-gyp-build')(__dirname);
    } catch (fallbackError) {
      throw new Error(
        `Failed to load native syslog module. This package only supports Linux x64/ARM64.\n` +
        `Original error: ${error instanceof Error ? error.message : String(error)}\n` +
        `Fallback error: ${fallbackError instanceof Error ? fallbackError.message : String(fallbackError)}`
      );
    }
  }
}

/**
 * Main Syslog class providing fluent interface for system logging
 * 
 * This class wraps the native syslog(3) system calls with a modern TypeScript API.
 * All methods are synchronous and return the instance for method chaining.
 * 
 * @example
 * ```typescript
 * const logger = new Syslog({
 *   ident: 'my-app',
 *   facility: Facilities.LOG_LOCAL0
 * });
 * 
 * logger.info('Application started')
 *       .error('Something went wrong', { code: 500 });
 * ```
 */
export class Syslog {
  private native: any;
  private initialized: boolean = false;
  private ident: string;
  private facility: SyslogFacility;
  private options: SyslogOption;

  /**
   * Create a new Syslog instance
   * 
   * Initializes the syslog connection with the provided options.
   * If no options are provided, sensible defaults are used.
   * 
   * @param options - Configuration options for the syslog connection
   * 
   * @example
   * ```typescript
   * // Default configuration
   * const logger = new Syslog();
   * 
   * // Custom configuration
   * const logger = new Syslog({
   *   ident: 'web-server',
   *   facility: Facilities.LOG_DAEMON,
   *   options: Options.LOG_PID | Options.LOG_CONS
   * });
   * ```
   */
  constructor(options: SyslogOptions = {}) {
    this.native = loadNativeModule();
    this.ident = options.ident || (typeof require !== 'undefined' && (require as any).main?.filename) || 'node';
    this.facility = options.facility ?? SyslogFacility.LOG_USER;
    this.options = options.options ?? SyslogOption.LOG_PID;
    
    this.initialize();
  }

  private initialize(): void {
    if (!this.initialized) {
      this.native.openlog(this.ident, this.options, this.facility);
      this.initialized = true;
    }
  }

  /**
   * Log an emergency message (LOG_EMERG)
   * 
   * Emergency level indicates system is unusable (level 0).
   * 
   * @param message - The message to log
   * @param context - Optional context object with additional data
   * @returns The Syslog instance for method chaining
   * 
   * @example
   * ```typescript
   * logger.emergency('System crash detected', { 
   *   error: 'kernel_panic',
   *   timestamp: Date.now()
   * });
   * ```
   */
  emergency(message: string, context?: Record<string, any>): this {
    this.log(SyslogLevel.LOG_EMERG, message, context);
    return this;
  }

  /**
   * Log an alert message (LOG_ALERT)
   * @param message Message to log
   * @param context Optional context object
   */
  alert(message: string, context?: Record<string, any>): this {
    this.log(SyslogLevel.LOG_ALERT, message, context);
    return this;
  }

  /**
   * Log a critical message (LOG_CRIT)
   * @param message Message to log
   * @param context Optional context object
   */
  critical(message: string, context?: Record<string, any>): this {
    this.log(SyslogLevel.LOG_CRIT, message, context);
    return this;
  }

  /**
   * Log an error message (LOG_ERR)
   * @param message Message to log
   * @param context Optional context object
   */
  error(message: string, context?: Record<string, any>): this {
    this.log(SyslogLevel.LOG_ERR, message, context);
    return this;
  }

  /**
   * Log a warning message (LOG_WARNING)
   * @param message Message to log
   * @param context Optional context object
   */
  warning(message: string, context?: Record<string, any>): this {
    this.log(SyslogLevel.LOG_WARNING, message, context);
    return this;
  }

  /**
   * Log a notice message (LOG_NOTICE)
   * @param message Message to log
   * @param context Optional context object
   */
  notice(message: string, context?: Record<string, any>): this {
    this.log(SyslogLevel.LOG_NOTICE, message, context);
    return this;
  }

  /**
   * Log an info message (LOG_INFO)
   * @param message Message to log
   * @param context Optional context object
   */
  info(message: string, context?: Record<string, any>): this {
    this.log(SyslogLevel.LOG_INFO, message, context);
    return this;
  }

  /**
   * Log a debug message (LOG_DEBUG)
   * @param message Message to log
   * @param context Optional context object
   */
  debug(message: string, context?: Record<string, any>): this {
    this.log(SyslogLevel.LOG_DEBUG, message, context);
    return this;
  }

  /**
   * Log a message at the specified level
   * @param level Syslog level
   * @param message Message to log
   * @param context Optional context object
   */
  log(level: SyslogLevel, message: string, context?: Record<string, any>): this {
    if (!this.initialized) {
      this.initialize();
    }

    let formattedMessage = message;
    
    if (context) {
      try {
        const contextStr = Object.entries(context)
          .map(([key, value]) => `${key}=${JSON.stringify(value)}`)
          .join(' ');
        formattedMessage = `${message} ${contextStr}`;
      } catch (error) {
        // If JSON.stringify fails, just use string representation
        const contextStr = Object.entries(context)
          .map(([key, value]) => `${key}=${String(value)}`)
          .join(' ');
        formattedMessage = `${message} ${contextStr}`;
      }
    }

    this.native.syslog(level, formattedMessage);
    return this;
  }

  /**
   * Close the syslog connection and cleanup resources
   * 
   * Calls the underlying closelog(3) system function.
   * After calling this, the instance can be reinitialized by
   * calling any logging method.
   * 
   * @example
   * ```typescript
   * logger.close();
   * // Can be reused:
   * logger.info('Reinitialized');
   * ```
   */
  close(): void {
    if (this.initialized) {
      this.native.closelog();
      this.initialized = false;
    }
  }

  /**
   * Get the current facility
   */
  getFacility(): SyslogFacility {
    return this.facility;
  }

  /**
   * Get the current ident
   */
  getIdent(): string {
    return this.ident;
  }

  /**
   * Get the current options
   */
  getOptions(): SyslogOption {
    return this.options;
  }

  /**
   * Check if the syslog connection is initialized
   */
  isInitialized(): boolean {
    return this.initialized;
  }
}

/**
 * Static access to facility constants
 */
export const Facilities: Record<string, number> = {
  LOG_KERN: SyslogFacility.LOG_KERN,
  LOG_USER: SyslogFacility.LOG_USER,
  LOG_MAIL: SyslogFacility.LOG_MAIL,
  LOG_DAEMON: SyslogFacility.LOG_DAEMON,
  LOG_AUTH: SyslogFacility.LOG_AUTH,
  LOG_SYSLOG: SyslogFacility.LOG_SYSLOG,
  LOG_LPR: SyslogFacility.LOG_LPR,
  LOG_NEWS: SyslogFacility.LOG_NEWS,
  LOG_UUCP: SyslogFacility.LOG_UUCP,
  LOG_CRON: SyslogFacility.LOG_CRON,
  LOG_AUTHPRIV: SyslogFacility.LOG_AUTHPRIV,
  LOG_FTP: SyslogFacility.LOG_FTP,
  LOG_LOCAL0: SyslogFacility.LOG_LOCAL0,
  LOG_LOCAL1: SyslogFacility.LOG_LOCAL1,
  LOG_LOCAL2: SyslogFacility.LOG_LOCAL2,
  LOG_LOCAL3: SyslogFacility.LOG_LOCAL3,
  LOG_LOCAL4: SyslogFacility.LOG_LOCAL4,
  LOG_LOCAL5: SyslogFacility.LOG_LOCAL5,
  LOG_LOCAL6: SyslogFacility.LOG_LOCAL6,
  LOG_LOCAL7: SyslogFacility.LOG_LOCAL7,
};

/**
 * Static access to level constants
 */
export const Levels: Record<string, number> = {
  LOG_EMERG: SyslogLevel.LOG_EMERG,
  LOG_ALERT: SyslogLevel.LOG_ALERT,
  LOG_CRIT: SyslogLevel.LOG_CRIT,
  LOG_ERR: SyslogLevel.LOG_ERR,
  LOG_WARNING: SyslogLevel.LOG_WARNING,
  LOG_NOTICE: SyslogLevel.LOG_NOTICE,
  LOG_INFO: SyslogLevel.LOG_INFO,
  LOG_DEBUG: SyslogLevel.LOG_DEBUG,
};

/**
 * Static access to option constants
 */
export const Options: Record<string, number> = {
  LOG_PID: SyslogOption.LOG_PID,
  LOG_CONS: SyslogOption.LOG_CONS,
  LOG_ODELAY: SyslogOption.LOG_ODELAY,
  LOG_NDELAY: SyslogOption.LOG_NDELAY,
  LOG_NOWAIT: SyslogOption.LOG_NOWAIT,
  LOG_PERROR: SyslogOption.LOG_PERROR,
};

/**
 * Create a new Syslog instance with default settings
 */
export function createSyslog(options?: SyslogOptions): Syslog {
  return new Syslog(options);
}

/**
 * Default syslog instance
 */
export const defaultLogger = new Syslog();

// Export convenience functions using the default logger
export const emergency = (message: string, context?: Record<string, any>) => 
  defaultLogger.emergency(message, context);

export const alert = (message: string, context?: Record<string, any>) => 
  defaultLogger.alert(message, context);

export const critical = (message: string, context?: Record<string, any>) => 
  defaultLogger.critical(message, context);

export const error = (message: string, context?: Record<string, any>) => 
  defaultLogger.error(message, context);

export const warning = (message: string, context?: Record<string, any>) => 
  defaultLogger.warning(message, context);

export const notice = (message: string, context?: Record<string, any>) => 
  defaultLogger.notice(message, context);

export const info = (message: string, context?: Record<string, any>) => 
  defaultLogger.info(message, context);

export const debug = (message: string, context?: Record<string, any>) => 
  defaultLogger.debug(message, context);

File: src/native/syslog.cpp
#include <napi.h>
#include <syslog.h>
#include <string.h>
#include <stdlib.h>

class SyslogNative : public Napi::ObjectWrap<SyslogNative> {
public:
    static Napi::Object Init(Napi::Env env, Napi::Object exports);
    SyslogNative(const Napi::CallbackInfo& info);

private:
    static Napi::FunctionReference constructor;
    
    Napi::Value Openlog(const Napi::CallbackInfo& info);
    Napi::Value Syslog(const Napi::CallbackInfo& info);
    Napi::Value Closelog(const Napi::CallbackInfo& info);
    Napi::Value GetFacilities(const Napi::CallbackInfo& info);
    Napi::Value GetLevels(const Napi::CallbackInfo& info);
    Napi::Value GetOptions(const Napi::CallbackInfo& info);
};

Napi::FunctionReference SyslogNative::constructor;

Napi::Object SyslogNative::Init(Napi::Env env, Napi::Object exports) {
    Napi::HandleScope scope(env);

    Napi::Function func = DefineClass(env, "SyslogNative", {
        InstanceMethod("openlog", &SyslogNative::Openlog),
        InstanceMethod("syslog", &SyslogNative::Syslog),
        InstanceMethod("closelog", &SyslogNative::Closelog),
        StaticMethod("getFacilities", &SyslogNative::GetFacilities),
        StaticMethod("getLevels", &SyslogNative::GetLevels),
        StaticMethod("getOptions", &SyslogNative::GetOptions),
    });

    constructor = Napi::Persistent(func);
    constructor.SuppressDestruct();

    exports.Set("SyslogNative", func);
    return exports;
}

SyslogNative::SyslogNative(const Napi::CallbackInfo& info) : Napi::ObjectWrap<SyslogNative>(info) {
    // Constructor logic if needed
}

Napi::Value SyslogNative::Openlog(const Napi::CallbackInfo& info) {
    Napi::Env env = info.Env();
    
    if (info.Length() < 3) {
        Napi::TypeError::New(env, "Expected 3 arguments: ident, option, facility")
            .ThrowAsJavaScriptException();
        return env.Null();
    }
    
    if (!info[0].IsString() || !info[1].IsNumber() || !info[2].IsNumber()) {
        Napi::TypeError::New(env, "Wrong argument types").ThrowAsJavaScriptException();
        return env.Null();
    }
    
    std::string ident = info[0].As<Napi::String>().Utf8Value();
    int32_t option = info[1].As<Napi::Number>().Int32Value();
    int32_t facility = info[2].As<Napi::Number>().Int32Value();
    
    openlog(ident.c_str(), option, facility);
    
    return env.Null();
}

Napi::Value SyslogNative::Syslog(const Napi::CallbackInfo& info) {
    Napi::Env env = info.Env();
    
    if (info.Length() < 2) {
        Napi::TypeError::New(env, "Expected 2 arguments: priority, message")
            .ThrowAsJavaScriptException();
        return env.Null();
    }
    
    if (!info[0].IsNumber() || !info[1].IsString()) {
        Napi::TypeError::New(env, "Wrong argument types").ThrowAsJavaScriptException();
        return env.Null();
    }
    
    int32_t priority = info[0].As<Napi::Number>().Int32Value();
    std::string message = info[1].As<Napi::String>().Utf8Value();
    
    syslog(priority, "%s", message.c_str());
    
    return env.Null();
}

Napi::Value SyslogNative::Closelog(const Napi::CallbackInfo& info) {
    Napi::Env env = info.Env();
    closelog();
    return env.Null();
}

Napi::Value SyslogNative::GetFacilities(const Napi::CallbackInfo& info) {
    Napi::Env env = info.Env();
    Napi::Object facilities = Napi::Object::New(env);
    
    facilities.Set("LOG_KERN", Napi::Number::New(env, LOG_KERN));
    facilities.Set("LOG_USER", Napi::Number::New(env, LOG_USER));
    facilities.Set("LOG_MAIL", Napi::Number::New(env, LOG_MAIL));
    facilities.Set("LOG_DAEMON", Napi::Number::New(env, LOG_DAEMON));
    facilities.Set("LOG_AUTH", Napi::Number::New(env, LOG_AUTH));
    facilities.Set("LOG_SYSLOG", Napi::Number::New(env, LOG_SYSLOG));
    facilities.Set("LOG_LPR", Napi::Number::New(env, LOG_LPR));
    facilities.Set("LOG_NEWS", Napi::Number::New(env, LOG_NEWS));
    facilities.Set("LOG_UUCP", Napi::Number::New(env, LOG_UUCP));
    facilities.Set("LOG_CRON", Napi::Number::New(env, LOG_CRON));
    facilities.Set("LOG_AUTHPRIV", Napi::Number::New(env, LOG_AUTHPRIV));
    facilities.Set("LOG_FTP", Napi::Number::New(env, LOG_FTP));
    facilities.Set("LOG_LOCAL0", Napi::Number::New(env, LOG_LOCAL0));
    facilities.Set("LOG_LOCAL1", Napi::Number::New(env, LOG_LOCAL1));
    facilities.Set("LOG_LOCAL2", Napi::Number::New(env, LOG_LOCAL2));
    facilities.Set("LOG_LOCAL3", Napi::Number::New(env, LOG_LOCAL3));
    facilities.Set("LOG_LOCAL4", Napi::Number::New(env, LOG_LOCAL4));
    facilities.Set("LOG_LOCAL5", Napi::Number::New(env, LOG_LOCAL5));
    facilities.Set("LOG_LOCAL6", Napi::Number::New(env, LOG_LOCAL6));
    facilities.Set("LOG_LOCAL7", Napi::Number::New(env, LOG_LOCAL7));
    
    return facilities;
}

Napi::Value SyslogNative::GetLevels(const Napi::CallbackInfo& info) {
    Napi::Env env = info.Env();
    Napi::Object levels = Napi::Object::New(env);
    
    levels.Set("LOG_EMERG", Napi::Number::New(env, LOG_EMERG));
    levels.Set("LOG_ALERT", Napi::Number::New(env, LOG_ALERT));
    levels.Set("LOG_CRIT", Napi::Number::New(env, LOG_CRIT));
    levels.Set("LOG_ERR", Napi::Number::New(env, LOG_ERR));
    levels.Set("LOG_WARNING", Napi::Number::New(env, LOG_WARNING));
    levels.Set("LOG_NOTICE", Napi::Number::New(env, LOG_NOTICE));
    levels.Set("LOG_INFO", Napi::Number::New(env, LOG_INFO));
    levels.Set("LOG_DEBUG", Napi::Number::New(env, LOG_DEBUG));
    
    return levels;
}

Napi::Value SyslogNative::GetOptions(const Napi::CallbackInfo& info) {
    Napi::Env env = info.Env();
    Napi::Object options = Napi::Object::New(env);
    
    options.Set("LOG_PID", Napi::Number::New(env, LOG_PID));
    options.Set("LOG_CONS", Napi::Number::New(env, LOG_CONS));
    options.Set("LOG_ODELAY", Napi::Number::New(env, LOG_ODELAY));
    options.Set("LOG_NDELAY", Napi::Number::New(env, LOG_NDELAY));
    options.Set("LOG_NOWAIT", Napi::Number::New(env, LOG_NOWAIT));
    options.Set("LOG_PERROR", Napi::Number::New(env, LOG_PERROR));
    
    return options;
}

Napi::Object Init(Napi::Env env, Napi::Object exports) {
    return SyslogNative::Init(env, exports);
}

NODE_API_MODULE(syslog_native, Init)